### PR TITLE
[Merged by Bors] - feat Port/Computability.RegularExpressions

### PR DESCRIPTION
**Pull request by arienmalec**
port of computability.regular_expressions

The previous file defined `attribute [match_pattern] Mul.mul` but this is not allowed in Lean 4

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by arienmalec** [comment]
Issues here are with `simpNF` & documented at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simpNF.20issues.20in.20Computability.2ERegularExpressions.20!4.232306/near/328277898

**Comment by arienmalec** [comment]
> Issues here are with `simpNF` & documented at https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simpNF.20issues.20in.20Computability.2ERegularExpressions.20!4.232306/near/328277898

now resolved.

**Comment by arienmalec** [comment]
Merging took this from green to red & I'm assuming the changes to `Language` was the issue.

Removing `map` from the failing `simp only` makes the `simp` pass, but now we have issues with termination

**Comment by arienmalec** [comment]
> Merging took this from green to red & I'm assuming the changes to `Language` was the issue.
> 
> Removing `map` from the failing `simp only` makes the `simp` pass, but now we have issues with termination

May be fixed when https://github.com/leanprover-community/mathlib/pull/18451 lands back on Mathlib4

File: Mathlib/Computability/RegularExpressions.lean [hunk 1]
```diff
@@ -95,71 +106,74 @@ theorem comp_def (P Q : RegularExpression α) : comp P Q = P * Q :=
...
-  | Char _ => false
+  | char _ => false
   | P + Q => P.matchEpsilon || Q.matchEpsilon
   | P * Q => P.matchEpsilon && Q.matchEpsilon
-  | star P => true
+  | star _P => true
 #align regular_expression.match_epsilon RegularExpression.matchEpsilon
 
-include dec
+--include dec
```
Comment (semorrison) [inline Mathlib/Computability/RegularExpressions.lean:169]:
```suggestion
```

File: Mathlib/Computability/RegularExpressions.lean [hunk 2]
```diff
@@ -308,21 +319,20 @@ theorem star_rmatch_iff (P : RegularExpression α) :
...
         · intro t' ht'
-          cases' ht' with ht' ht'
-          · rw [ht']
-            exact ⟨by decide, ht⟩
-          · exact helem _ ht'
-    · rintro ⟨S, hsum, helem⟩
+          cases' ht' with ht' ht' -- porting note: last ht' doesn't take, need to `rename_i` below
+          · simp only [ne_eq, not_false_iff, true_and, rmatch]
+            exact ht
+          · rename_i ht'; exact helem t' ht'
```
Comment (semorrison) [inline Mathlib/Computability/RegularExpressions.lean:325]: Can you make this work by using the structured cases syntax?

**Comment by semorrison** [comment]
Looks pretty close! Sorry this one has taken so long.

File: Mathlib/Computability/RegularExpressions.lean (hunk 2 above)
Comment (arienmalec) [reply Mathlib/Computability/RegularExpressions.lean]: works!

File: Mathlib/Computability/RegularExpressions.lean [hunk 3]
```diff
@@ -319,10 +319,14 @@ theorem star_rmatch_iff (P : RegularExpression α) :
         constructor
         · simp [hs, hsum]
         · intro t' ht'
-          cases' ht' with ht' ht' -- porting note: last ht' doesn't take, need to `rename_i` below
-          · simp only [ne_eq, not_false_iff, true_and, rmatch]
+          -- porting note: was `cases' with ht' ht'` but a `cases'` bug
+          -- caused the last `ht'` not to take, requiring `rename_i`
+          -- moved to explicit cases.
```
Comment (semorrison) [inline Mathlib/Computability/RegularExpressions.lean:323]: I think you can just remove the porting note, at this point. `cases'` is only intended as a stopgap measure, so once you've rewritten it using structured cases this file doesn't need revisiting.

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**